Challenge: Make Your Network

Create a multi-layer perceptron neural network model to predict on a labeled dataset of your choosing. Compare this model to either a boosted tree or a random forest model and describe the relative tradeoffs between complexity and accuracy. Be sure to vary the hyperparameters of your MLP!

Note, I am using the same red wine quality dataset that was used in my Supervised Learning Capstone for which the best performing model was the Random Forest Classifier (89.3% accuracy rate). There have been additional Q&A discussions regarding the potential for building a product based on the results from my Supervised Learning Capstone. Part of the discussion has been applying an Unsupervised Learning approach to see how accurately an algorithm like MLP would also classify the quality of wine based on physicochemical compounds, labeled data.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split

In [2]:
# Loading the dataset.
red_wine = pd.read_csv('C:\\Users\\kenne\\Desktop\\winequality-red.csv', encoding='utf-8')

In [4]:
# Binarize the target variable quality where 0 equals bad and 1 equals good.
grade = []
for qual in red_wine['quality']:
    if qual >= 7:
        grade.append(1)
    else:
        grade.append(0)
red_wine['grade'] = grade

red_wine.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality        grade  
count  1599.000000  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023     0.135710  
std       0.154386     0.169507     1.065668     0.807569     0.342587  
min       2.740000     0.330000     8.400000     3.000000     0.000000  
25%       3.210000     0.550000     9.500000     5.000000     0.000000  
50%       3.310000     0.620000    10.200000     6.000000     0.000000  
75%       3.400000     0.730000    11.100000     6.000000     0.000000  
max       4.010000     2.000000    14.900000     8.000000     1.000000

In [5]:
red_wine = red_wine.drop(['quality'], axis=1)

red_wine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  grade  
0      9.4      0  
1      9.8      0  
2      9.8      0  
3      9.8      0  
4      9.4      0

In [7]:
# Split the data into X and y variables.
X = red_wine.iloc[:, :11]
Y = red_wine['grade']

First attempt, Multi-layer Perceptron model (MLP) ...

In [9]:
# Establish and fit the model, with a single, 1000 perceptron layer.
mlp = MLPClassifier(hidden_layer_sizes=(1000))
mlp.fit(X, Y)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=1000, learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [10]:
mlp.score(X, Y)

0.8642901813633521

In [11]:
Y.value_counts()/len(Y)

0    0.86429
1    0.13571
Name: grade, dtype: float64

In [12]:
# Five-fold cross-valiation.
from sklearn.model_selection import cross_val_score
cross_val_score(mlp, X, Y, cv=5)

array([0.87850467, 0.86292835, 0.86520376, 0.84639498, 0.86833856])

Produced an accuracy rate of 86.4% which is slightly less than the best model from the Supervised Learning approach, Random Forest Classifier.

Cross-validation scores are consistent.

Second attempt, adjusting the layers ...

In [13]:
# Establish and fit the model, with a single, 500 perceptron layer.
mlpv2 = MLPClassifier(hidden_layer_sizes=(500))
mlpv2.fit(X, Y)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=500, learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [14]:
mlpv2.score(X, Y)

0.8642901813633521

In [15]:
Y.value_counts()/len(Y)

0    0.86429
1    0.13571
Name: grade, dtype: float64

In [16]:
# Five-fold cross-valiation.
from sklearn.model_selection import cross_val_score
cross_val_score(mlpv2, X, Y, cv=5)

array([0.87538941, 0.86292835, 0.86833856, 0.830721  , 0.86520376])

No improvement in accuracy rate, similar score to the 1000 layer attempt.

Cross-validation scores still consistent but not as consisent as previous attempt.

Third attempt, adjust the layers further ...

In [17]:
# Establish and fit the model, with double, 750 perceptron layer.
mlpv3 = MLPClassifier(hidden_layer_sizes=(750, 750))
mlpv3.fit(X, Y)

MLPClassifier(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
       beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(750, 750), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [18]:
mlpv3.score(X, Y)

0.8642901813633521

In [19]:
Y.value_counts()/len(Y)

0    0.86429
1    0.13571
Name: grade, dtype: float64

In [20]:
# Five-fold cross-valiation.
from sklearn.model_selection import cross_val_score
cross_val_score(mlpv3, X, Y, cv=5)

array([0.86604361, 0.83489097, 0.89028213, 0.80564263, 0.86520376])

No improvement in accuracy rate, similar score to both previous attempts.

Cross-validation scores not as consistent as previous attempts. Will try varying other hyperparameters now.

Fourth attempt, logistic activation ...

In [21]:
# Establish and fit the model, logistic activation.
mlpv4 = MLPClassifier(activation='logistic')
mlpv4.fit(X, Y)

MLPClassifier(activation='logistic', alpha=0.0001, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [22]:
mlpv4.score(X, Y)

0.8649155722326454

In [23]:
Y.value_counts()/len(Y)

0    0.86429
1    0.13571
Name: grade, dtype: float64

In [24]:
# Five-fold cross-valiation.
from sklearn.model_selection import cross_val_score
cross_val_score(mlpv4, X, Y, cv=5)

array([0.87538941, 0.83800623, 0.87460815, 0.84326019, 0.86520376])

We finally have some improvement but it is very small, too small to matter (percentages between good and bad wine did not change).

Cross-validation scores are relatively consistent.

Fifth attempt, reducing alpha ...

In [25]:
# Establish and fit the model, with default settings.
mlpv5 = MLPClassifier(activation='logistic', hidden_layer_sizes=(1000), alpha=1e-6)
mlpv5.fit(X, Y)

MLPClassifier(activation='logistic', alpha=1e-06, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=1000, learning_rate='constant',
       learning_rate_init=0.001, max_iter=200, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=None,
       shuffle=True, solver='adam', tol=0.0001, validation_fraction=0.1,
       verbose=False, warm_start=False)

In [26]:
mlpv5.score(X, Y)

0.8792995622263915

In [27]:
Y.value_counts()/len(Y)

0    0.86429
1    0.13571
Name: grade, dtype: float64

In [28]:
# Five-fold cross-valiation.
from sklearn.model_selection import cross_val_score
cross_val_score(mlpv5, X, Y, cv=5)

array([0.86604361, 0.85358255, 0.86520376, 0.80877743, 0.86520376])

Accuracy rate increased by a noticeable amount!

Cross-validation scores were relatively consisent, with the exception of one score.

Overall, the Random Forest Classifier still performed better than any variation of a Multi-Layer Perceptron model.